src/Npgsql/Internal/Converters/RecordObjectArrayConverter.cs

```diff
@@ -0,0 +1,71 @@
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Npgsql.Internal.Descriptors;
+using Npgsql.PostgresTypes;
+using Npgsql.Properties;
+
+namespace Npgsql.Internal.Converters;
+
+public class RecordObjectArrayConverter : PgStreamingConverter<object[]>
+{
+    readonly PgSerializerOptions _serializerOptions;
+
+    public RecordObjectArrayConverter(PgSerializerOptions serializerOptions)
+        => _serializerOptions = serializerOptions;
+
+    public override object[] Read(PgReader reader)
+        => Read(async: false, reader, CancellationToken.None).GetAwaiter().GetResult();
+
+    public override ValueTask<object[]> ReadAsync(PgReader reader, CancellationToken cancellationToken = default)
+        => Read(async: true, reader, cancellationToken);
+
+    async ValueTask<object[]> Read(bool async, PgReader reader, CancellationToken cancellationToken)
+    {
+        if (reader.ShouldBuffer(sizeof(int)))
+            await reader.BufferData(async, sizeof(int), cancellationToken).ConfigureAwait(false);
+        var fieldCount = reader.ReadInt32();
+        var result = new object[fieldCount];
+
+        for (var i = 0; i < fieldCount; i++)
+        {
+            if (reader.ShouldBuffer(sizeof(uint) + sizeof(int)))
+                await reader.BufferData(async, sizeof(uint) + sizeof(int), cancellationToken).ConfigureAwait(false);
+
+            var typeOID = reader.ReadUInt32();
+            var fieldLen = reader.ReadInt32();
+
+            // Note that we leave .NET nulls in the object array rather than DBNull.
+            if (fieldLen == -1)
+                continue;
+
+            var postgresType =
```

Comment:
Hopefully this is all correct - take a look.

Reply:
General shape looks good to me, albeit a bit more verbose than was previously the case.

Probably want GetObjectOrDefault info here though, to prefer the object converter if it exists. (will for instance correctly handle arraynullability mode, if an array value is a field in this record)

Reply:
Switched from GetDefaultTypeInfo to GetObjectOrDefault.

Yeah, definitely more verbose... Not a huge deal, it matters only for compositional converters anyway.